agent/xds/clusters.go

```diff
@@ -0,0 +1,79 @@
+package xds
+
+import (
+	"errors"
+	"time"
+
+	envoy "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoyauth "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	envoycore "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/hashicorp/consul/agent/proxycfg"
+)
+
+// clustersFromSnapshot returns the xDS API representation of the "clusters"
+// (upstreams) in the snapshot.
+func clustersFromSnapshot(cfgSnap *proxycfg.ConfigSnapshot, token string) ([]proto.Message, error) {
+	if cfgSnap == nil {
+		return nil, errors.New("nil config given")
+	}
+	// Include the "app" cluster for the public listener
+	clusters := make([]proto.Message, len(cfgSnap.Proxy.Upstreams)+1)
+
+	clusters[0] = makeAppCluster(cfgSnap)
+
+	for idx, upstream := range cfgSnap.Proxy.Upstreams {
+		clusters[idx+1] = makeUpstreamCluster(upstream.Identifier(), cfgSnap)
+	}
+
+	return clusters, nil
+}
+
+func makeAppCluster(cfgSnap *proxycfg.ConfigSnapshot) *envoy.Cluster {
+	addr := cfgSnap.Proxy.LocalServiceAddress
+	if addr == "" {
+		addr = "127.0.0.1"
+	}
+	return &envoy.Cluster{
+		Name: LocalAppClusterName,
+		// TODO(banks): make this configurable from the proxy config
```

Comment:
TODO now or later?

Reply:
Post 1.3. I have a (paper) todo list to make issues for things like this for near-future iterations.